src/anndata/_core/xarray.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import pandas as pd
+
+from ..compat import XDataset
+
+if TYPE_CHECKING:
+    from ..compat import XDataArray
+
+
+class Dataset2D(XDataset):
+    """
+    A wrapper class meant to enable working with lazy dataframe data.
+    We do not guarantee the stability of this API beyond that guaranteed
+    by :class:`xarray.Dataset` and the `to_memory` function, a thin wrapper
+    around :meth:`xarray.Dataset.to_dataframe` to ensure roundtrip
+    compatibility here.
+    """
+
+    __slots__ = ()
+
+    @property
+    def is_backed(self) -> bool:
+        return self.attrs.get("is_backed", False)
+
+    @is_backed.setter
+    def is_backed(self, isbacked: bool):
+        if not isbacked and "is_backed" in self.attrs:
+            del self.attrs["is_backed"]
+        else:
+            self.attrs["is_backed"] = isbacked
+
+    @property
+    def index_dim(self) -> str:
+        if len(self.sizes) != 1:
+            msg = f"xarray Dataset should not have more than 1 dims, found {len(self.sizes)} {self.sizes}, {self}"
+            raise ValueError(msg)
+        return next(iter(self.coords.keys()))
+
+    @property
+    def true_index_dim(self) -> str:
+        index_dim = self.attrs.get("indexing_key", None)
+        return index_dim if index_dim is not None else self.index_dim
+
+    @true_index_dim.setter
+    def true_index_dim(self, val: str):
+        if val is None or (val == self.index_dim and "indexing_key" in self.attrs):
+            del self.attrs["indexing_key"]
+        elif val not in self.dims:
+            if val not in self.data_vars:
+                msg = f"Unknown variable `{val}`."
+                raise ValueError(msg)
+            self.attrs["indexing_key"] = val
+
+    @property
+    def xr_index(self) -> XDataArray:
+        return self[self.index_dim]
+
+    @property
+    def index(self) -> pd.Index:
+        """:attr:`~anndata.AnnData` internally looks for :attr:`~pandas.DataFrame.index` so this ensures usability
+
+        Returns
+        -------
+        The index of the of the dataframe as resolved from :attr:`~xarray.Dataset.coords`.
+        """
+        return self.indexes[self.index_dim]
+
+    @index.setter
+    def index(self, val) -> None:
+        index_dim = self.index_dim
+        self.coords[index_dim] = (index_dim, val)
+        if isinstance(val, pd.Index) and val.name is not None and val.name != index_dim:
+            self.update(self.rename({self.index_dim: val.name}))
+            del self.coords[index_dim]
+        if "indexing_key" in self.attrs:
+            del self.attrs["indexing_key"]
```

Comment:
Why do we delete the `indexing_key` in the `if` branch here, `if "indexing_key" in self.attrs:`?  Maybe should comment, at first glance I don't see why `index` and `true_index` (for which `indexing_key` is the key, not necessarily a dimension/coordinate) are related.

Reply:
I figured if someone is setting a new index on the object, the expectation is that this is the one definitive index without anything else. But I can remove that branch, I don't think that behavior is used anywhere at the moment.

Reply:
Ok, suggested a change above then, I see the logic a bit more now.
```suggestion
        index_dim = self.index_dim
        self.coords[index_dim] = (index_dim, val)
        if isinstance(val, pd.Index) and val.name is not None and val.name != index_dim:
            self.update(self.rename({self.index_dim: val.name}))
            del self.coords[index_dim]
        # without `indexing_key` explicitly set on `self.attrs`, `self.true_index_dim` will use the `self.index_dim`
        if "indexing_key" in self.attrs:
            del self.attrs["indexing_key"]
```